wxDVC accessibility tweaks

vslavik opened the pull request:
These improve behavior with JAWS in particular.

The `NotifyEvent` would be much nicer with C++11, of course, but... I'm also a bit worried about it, but it can always be reversed if it causes problems (FWIW, I didn't find any with JAWS or NVDA).

vslavik committed:
Don't use wxACC_STATE_SYSTEM_READONLY for wxDVC items

Items in wxDataViewCtrl are only read-only if there's no editable
column, in which case all items are read-only. This flag is only useful
for accessibility if only some of the items in the list are read-only;
if all of them are, it's just noise.

vslavik committed:
Defer wxAccessible::NotifyEvent() to idle time

Change NotifyEvent() implementation so that it doesn't call
::NotifyWinEvent() immediately, but only in the next iteration of the
event loop. This makes it easier to write code that uses, because it
must be called after the changes it notifies about were fully made and
are visible in the corresponding window's accessible interface.

vadz commented on src/msw/ole/access.cpp:
```diff
@@ -1827,8 +1827,28 @@ IAccessible *wxAccessible::GetIAccessibleStd()
+
+    // send the notification in idle time to be sure it is sent after the change
+    // was fully done in wx code
+    struct SendNotification
```
I think this is not going to compile with g++ in non-C++11 mode, template parameters must have extern linkage in C++98.

IOW this struct should be moved out of the function.

vadz reviewed (approved):
No real comments, just as you, I'm not *sure* this is not going to result in any problems, but let's at least experiment with it -- hopefully we'll get more testing of accessibility features now that they actually work for this control.

vslavik commented:
> hopefully we'll get more testing of accessibility features now that they actually work for this control.

That's debatable, `wxUSE_ACCESSIBILITY` is still off by default and its previous state implies almost nobody enables it. However, now that it's much more lightweight and doesn't normally interfere with Windows' default accessibility implementation, I think it makes sense to enable it by default — should I?

vadz commented:
Oops, sorry, I was somehow sure that you had already enabled `wxUSE_ACCESSIBILITY` by default now, which was why I thought we'd get more feedback about this code now.

And yes, AFAICS, we really should enable it now as there doesn't seem to be any drawbacks to doing it, are there any I'm missing?

vslavik committed:
Enable wxUSE_ACCESSIBILITY by default (MSW)

Compile accessibility support on Windows by default now that the generic
wxDataViewCtrl control implements accessible interface. After the
changes from 7dab555f7118f24e70444ff6d6411433b730b889, accessibility
support is much more lightweight and doesn't interfere with normal win32
behavior, so this change shouldn't affect accessibility-unaware code in
any way.

vslavik commented:
> And yes, AFAICS, we really should enable it now as there doesn't seem to be any drawbacks to doing it, are there any I'm missing?

Not, just a small size increase.

vslavik committed:
Add virtual destructor to wxIAccessible

Silence gcc warning about non-virtual destructor.
